impy/models/sophia.py

```diff
@@ -5,158 +5,191 @@
 '''
 
 import numpy as np
-from impy.common import standard_particles
-
-
-class SophiaCascadeRun():
-    def __init__(self,
-                 lib_str,
-                 label,
-                 decay_mode,
-                 n_events,
-                 fill_subset=False,
-                 p_debug=False,
-                 nucleon_Ekin=None):
-        from ParticleDataTool import SibyllParticleTable
-        exec("import " + lib_str + " as siblib")
-        self.lib = siblib  # @UndefinedVariable
-        self.label = label
-        self.nEvents = n_events
-        self.fill_subset = fill_subset
-        self.spectrum_hists = []
-        self.dbg = p_debug
-        self.nucleon_Ekin = nucleon_Ekin
-        set_stable(self.lib, decay_mode, self.dbg)
-        self.ptab = SibyllParticleTable()
-        self.init_generator()
-
-    def init_generator(self):
+from impy.common import MCRun, MCEvent, impy_config
+from impy.util import info
+
+
+
+class SophiaEvent(MCEvent):
+    """Wrapper class around Sophia code"""
+
+    def __init__(self, lib, event_kinematics, event_frame):   
+
+        lib.toevt() # prepare hepevt common block     
+        np = lib.hepevt.nhep # number of particles in event
+        pem_arr = lib.s_plist.p[:np,:].T # array of particle momenta
+        vt_arr = lib.hepevt.vhep[:, 0:np] # array of verticies (x, y, z, t) - all zeros for sophia
+        
+        MCEvent.__init__(self,
+                         lib = lib,
+                         event_kinematics = event_kinematics,
+                         event_frame = event_frame,
+                         nevent = lib.hepevt.nevhep, # event number
+                         npart = np,
+                         p_ids = lib.hepevt.idhep[0:np],
+                         status = lib.hepevt.isthep[0:np],
+                         px = pem_arr[0],
+                         py = pem_arr[1],
+                         pz = pem_arr[2],
+                         en = pem_arr[3],
+                         m = pem_arr[4],
+                         vx = vt_arr[0],
+                         vy = vt_arr[1],
+                         vz = vt_arr[2],
+                         vt = vt_arr[3],
+                         pem_arr = pem_arr,
+                         vt_arr = vt_arr)
+
+    def filter_final_state(self):
+        self.selection = np.where(self.status == 1)
+        self._apply_slicing()
+
+    def filter_final_state_charged(self):
+        self.selection = np.where((self.status == 1) & (self.charge != 0))
+        self._apply_slicing()
+    
+    @property
+    def charge(self):
+        return self.lib.schg.ichg[self.selection]
+    
+    @property
+    def decayed_parent(self):
+        """Returns the array of indices of the decayed parent particles"""
+        MCEvent.parents(self)
+        return self.lib.schg.decpar[0:self.npart]
+
+    @property
+    def parents(self):
+        """In SOPHIA parents are difficult to obtain. This function returns 0."""
```

Comment:
Isn't this working? Then should be reflected in the docstring. Also for the children below.

Reply:
It returns an array of correct shape with all zeros. I added this info to the docstring.